enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+  will be performed with MicroShift cluster not running.
+- To allow database upgrades only etcd and kube-apiserver will be active during
+  data migration.
+- System start was chosen as point in time that above actions will happen,
+  just before starting MicroShift cluster.
+
+### Workflow Description
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+Upgrade:
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots new commit
+1. Operating System, greenboot, and MicroShift take actions (migrating database
+   content, causing a rollback, etc.) without any additional intervention
+
+Manual rollback:
+
+1. MicroShift administrator rollbacks to or stages an ostree commit with MicroShift
+   that was already running on the device and performs a reboot
+1. Staged ostree commit boots
+1. MicroShift will restore the backup matching current ostree commit
+1. MicroShift will run.
+
+### API Extensions
+
+Metadata persisted on filesystem related to the functionality described in this enhancement
+is considered internal implementation detail and not an API intended for user.
+Schemas and locations of these files are subject to change.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Being a GA feature from the beginning the risks are not foreseeing fail scenarios in advance and implementation bugs
+that are not caught and fixed through graduation process.
+
+To mitigate the risks, a thorough review of the enhancement must be done by MicroShift, OpenShift, and RHEL teams,
+and making sure testing strategy is sound and prioritized equally with the feature development.
+
+### Drawbacks
+
+N/A
+
+## Design Details
+
+### Definitions
+
+- **ostree commit**: image containing root filesystem
+- **ostree deployment**: bootloader entries created from ostree commits
+  (this document refers to "system image" very loosely as both "commit" and "deployment")
+- **Rollback**: booting older (that already ran on the device) ostree commit -
+  either due to greenboot or manual intervention
+- **Backup**: backing up MicroShift's data
+- **Restore**: restoring MicroShift's data from a backup
+- **Data migration**: procedure of transitioning MicroShift's data to be compatible with newer binary
+- **Version metadata**: file storing MicroShift's version and ostree commit ID
+- **MicroShift greenboot healthcheck**: program verifying the status of MicroShift's cluster
+
+### Phases of execution
+
+1. Pre run phase
+   - Failure blocks start of MicroShift's cluster
+   - Backs up or restores data
+   - Migrates data to newer schemas if needed
+1. Run phase
+   - Start of MicroShift's cluster
+
+In parallel:
+1. greenboot runs MicroShift health check
+1. greenboot runs red or green scripts depending on system's health
```

Comment:
microshift specific red/green scripts will be shipped with microshift (microshift-greenboot rpm specifically), but they just invoke `microshift` binary commands
these commands will persist info to `/var` which is the mutable part of the filesystem (doesn't change when different deployment is started)

Reply:
i'm trying to understand the implications for:

a) microshift ships buggy health scripts so they are themselves the reason the upgrade fails (does rolling back "fix" the health scripts?)
b) microshift ships health scripts that aren't version compatible to n-1 or n+1 (does it matter or because they're paired to the microshift version and upgraded/rolledback with microshift itself, they only after need to be compatible with one microshift version)

Reply:
> a) microshift ships buggy health scripts so they are themselves the reason the upgrade fails (does rolling back "fix" the health scripts?)
yes, healthchecks are shipped with MicroShift, so ostree commit with different MicroShift will result in different healthchecks
they reside in `/etc/greenboot/` and `/etc` is in the immutable part of the filesystem

> b) microshift ships health scripts that aren't version compatible to n-1 or n+1 (does it matter or because they're paired to the microshift version and upgraded/rolledback with microshift itself, they only after need to be compatible with one microshift version)
I think ^ answers this - new MicroShift RPM can bring new healthcheck scripts.

Currently our healthcheck script is written in bash, but (as an optimization) we might rewrite it to go and make part of `microshift` executable so it would be "always" up to date (would know what specific "system" deployments, daemonsets, etc. it should expect and verify that)

Reply:
yeah i think that answers my question.  they are tied to the microshift version, so:

1) rolling back microshift should roll you back to previously working/valid health check scripts
2) they don't need to be cross-version compatible since they are included with microshift itself.

it might be worth documenting that in the EP though.

Reply:
Added